### ISSUE TITLE
JuliaCon submission comments

### ISSUE THREAD

**Issue by snoeyink** [avik-pal/RayTracer.jl#9]
Here are some comments on your JuliaCon submission in pdf.

The main things are:
1. Title and abstract mislead me about the paper contents, so I have suggestions. Feel free to use your own rewordings, since I may have misplaced the emphasis. 
2. Experiments need enough detail that they can be replicated, esp for BVH and light positioning.
3. Plots in fig 1 & 2 should factor out the expected trend and be single column. 
4. Little, but important details: drop the brackets around figure & algorithm numbers (makes them look like paper citations.), and keep the aspect ratio from Fig 4 a,b in c,d.
5. Comments on writing, particularly in consistent use of terms and context first.  See Gopen & Swan's https://cseweb.ucsd.edu/~swanson/papers/science-of-writing.pdf
6. The issue of ray tracing being non-differentiable is mentioned in the first paragraph of the intro, but you never say how that affects what you can attempt.  Good science points out the limitations of a technique. 

See attached pdfs
[Comments by page on 10.21105.jcon.00031 (1).pdf](https://github.com/avik-pal/RayTracer.jl/files/3374859/Comments.by.page.on.10.21105.jcon.00031.1.pdf)
[Comments on 10.21105.jcon.00031 (1).pdf](https://github.com/avik-pal/RayTracer.jl/files/3374860/Comments.on.10.21105.jcon.00031.1.pdf)

**Comment by avik-pal** [comment]
@snoeyink Thank you for the detailed review. I am working on an updated version of the paper keeping all the points in mind.

> What is the support for this claim? How does the non-differentiability of model-building decisions limit it?

I am adding a current limitations section to the paper where I shall explain the cases where the RayTracer should not be used and why we did not handle those cases explicitly.

> Why is the number of pixels the measure? I'd expect it to be the number of mesh elements. 

The main issue here is to measure the time/memory in a way that it stays consistent with increasing the mesh elements. If we increase the number of elements in the mesh it is not necessary that the time to render it will increase. It will be dependent on the configuration of the mesh (so the dependency would not only be the no of mesh elements but a variety of other factors). Hence, the only way I could consistently increase the complexity by a single parameter was to increase the no of pixels keeping the mesh constant.

> How fast is the gradient computation after AD? (How fast is the AD?) How do these grow with scene complexity? 

I will put a comparison between the performance of AD vs Numerical Gradients.

**Comment by dourouc05** [comment]
A few other comments (I build upon what @snoeyink said and thus not repeating too much). I really like the approach and your use cases! That's really something that more people should know about, and more software like this is a great step in that direction. 

Major comments: 

* Did you perform any hyperparameter optimisation? (Like learning rate, parameters for Adam.) If so, how? 
* Overall, your submission felt a bit lacking: no performance figures, no real evaluation of the optimisation algorithms (Adam is not the only one) or of the hyperparameters; also, you don't tell a word about Duckietown, although I think this is a very practical use case. 
* "We interface our renderer with the deep learning library Flux for use in combination with neural networks": what combinations are you making with neural networks? This is the only mention of neural networks in the whole submission I can find. 
* You say that automatic differentiation is not so available in languages like C++, but there are packages like CppAD (https://github.com/coin-or/CppAD) or CasADi (https://web.casadi.org/): you should mention them and discuss why they are not applicable. 
* How do you compare to redner? You just mention its existence, but do not compare yourself to it. Your package doesn't really have to shine, but that would really be useful information to user to decide which library they should use/invest on (functionalities, performance, ease of extension, etc.). 

Minor ones: 

* Figure 1 is not very legible, you seem to have stretched the image to make it fit in the page. (The same is true, albeit to a lesser extent, for Figures 3 and 4. For Figure 4, try to have all plot titles have the same size.)
* "For these measurements we use a mesh with 137 centered": a word is missing (137 what?). 
* In Algorithm 1, I don't understand your syntax "gs = gradient(params) do". Are you just defining a variable `gs`? If so, why the `do-end`? 
* I would add more context to the abstract, to briefly explain a potential use case (within the limits of JuliaCon). 

About the software: 

* The constructor for `Camera` should use keyword arguments, because as is the code is hard to read (like in your Listing 1): there are too many arguments for position to be enough. The same goes for the function `raytrace`. It wouldn't hurt for PointLight, but it feels less necessary. 
* I don't think the package name is well-suited: you might want to have other ray-tracers in Julia, with other trade-offs (better performance but lack of differentiability, for instance). 
* For your test suite, I think that a test like https://github.com/avik-pal/RayTracer.jl/blob/master/test/mesh_render.jl could be improved upon a lot: why don't you check the return values? You could have reference images and compare them to the results you get from the code. This is slightly brittle, but that's often the case with integration tests (and I've seen this approach for GUI toolkits). You could remain at the unit test by ensuring that the correct graph is built before being passed to AD, but that would be more difficult to write and, in my opinion, less useful. Otherwise, you really should do a lot more unit testing: testing small bits of your code, each and every public function, so that you can easily spot which function stopped working. 
* I feel that your documentation is not polished enough: it is quite good for a prototype, but not enough for software to be published and used by outside people. For instance, in https://avik-pal.github.io/RayTracer.jl/dev/getting_started/rendering/#A-Step-By-Step-Introduction-to-Rendering-1, you give a lot of code, but you don't really explain the parameters you give to each function (you should also have links to the "reference" part of the documentation). 
* In your submission, you speak about Optim.jl, but I did not see anything related to that package in your code: is there a reason for this? 
* You don't really explain why you need Julia 1.1 (might be worth a small note, because Julia 1.0 is the long-term version and thus may be very relevant for some people).

**Comment by avik-pal** [comment]
@dourouc05 Thank you for your feedback.

## Software Concerns

I have tried to address the software issues you mentioned in #13 
1. Now `raytrace`, `Camera` and `Light` have keyword arguments. But the original API also exists.
2. All the functions are now documented with some detailed getting started tutorials.
3. I have added a tutorial on how to use Optim for scene reconstruction
4. Added a note on why julia 1.0 cannot be used

I will deal with the remaining software Issues like the test suite in the next few days.

**Comment by avik-pal** [comment]
I will get back to the other comments once JuliaCon is over.

**Comment by dourouc05** [comment]
No problem, you already did a good job :+1:!

**Comment by snoeyink** [comment]
I too appreciate the changes that you made and think they are a definite improvement.  

Here are some typos that you missed. The experiment that generates Fig 3 could use more detail to make it repeatable.  [Summary of Comments on 10.21105.jcon.00037.pdf](https://github.com/avik-pal/RayTracer.jl/files/3418217/Summary.of.Comments.on.10.21105.jcon.00037.pdf)
[10.21105.jcon.00037.pdf](https://github.com/avik-pal/RayTracer.jl/files/3418218/10.21105.jcon.00037.pdf)

Enjoy JuliaCon.

**Comment by avik-pal** [comment]
Just a small note, even though I might have mentioned some urls in the paper which refer to the master branch, they are currently in the `review_updates`. In future I will list them in this issue separately. 
[https://github.com/avik-pal/RayTracer.jl/examples](https://github.com/avik-pal/RayTracer.jl/examples) --> [https://github.com/avik-pal/RayTracer.jl/tree/review_updates/examples](https://github.com/avik-pal/RayTracer.jl/tree/review_updates/examples)

**Comment by avik-pal** [comment]
> In Algorithm 1, I don't understand your syntax "gs = gradient(params) do". Are you just defining a variable gs? If so, why the do-end?

The `do-end` syntax simply transforms the function enclosed in it to the following form:
```
gs = gradient(params) do
     l1 = operation1(params)
     return operation2(l1)
end
```

```
function myfunction(s)
     l1 = operation1(s)
     return operation2(l1)
end
gs = gradient(myfunction, params)
```

**Comment by dourouc05** [comment]
I think it would be best to avoid this syntax in the article, as this syntax is quite specific to Julia.

**Comment by snoeyink** [comment]
I am completely fine with `do` syntax in a JuliaCon proceedings paper, although @dourouc05's comment suggests that it would be good to briefly explain it. 

Some comments on the latest revision, which continues to improve on the original submission. 
 
Figure 3 and the revised Fig 4 are nice additions.  
I did have to read the whole of section 5.2 before I can understand the context for the first paragraph of that section -- perhaps you can start with the question that you are trying to answer, rather than a statement "we compare the use".  (The best writers "[show, don't tell.](https://jerryjenkins.com/show-dont-tell/)"  Think about the reader's perspective, and get them to ask the question, "is AD better than numerical differentiation?"  just before you provide evidence for the answer.)

Section 2, I like: 
> Differentiation allows such renderers to make use of gradients to learn the inverse mapping from an image to its parameter space. 

But I think the rest of the paragraph needs evidence in this paper if that is going to be a prominent claim.  The devil is in the details, so a citation to [3] is wishful thinking or complete bullshit. 

Fig 2: I strongly suspect is that what the data shows is that the BVH takes half the time (and half the memory allocations, which I suspect is related).  You can't see that in at the left, but only at the right of each plot, since the plots are dominated by the exponential growth in number of pixels/parameters.  Please try plotting time per pixel/parameter (or sqrt(pixel or parameter)) rather than total time to see if this reveals if the multiplier is consistent at 1/2.

The review checklist suggests that DOI's should be included for references that have them.

**Comment by avik-pal** [comment]
I have added the DOI's in the latest revision

**Comment by dourouc05** [comment]
Some proofreading would help (most notably with commas: you use very few of them, which makes the text rather hard to read). Two typos in Section 5.5: "The converge of the model […] paramters" -> convergence and parameters. 

Figure 3b: the y-axis is not clear (acceleration factor?). Also, how good were the gradients computed by finite differences? Maybe you could give some insight into this method? Also, how do you choose the step to apply the formula (i.e. the h in (f(x-h) - f(x+h)/2h)? 

A bit of nit-picking: in Algo. 1, you never define the value of `iter`, nor do you update it. Similarly, `converged` is never set to `false`, and there is no explicit initial value for `params`.

**Comment by dourouc05** [comment]
The file `paper/paper.bbl` should not be present on the repo, as it is automatically generated from the BibTeX file (add it to the gitignore?).

For me, that's it :)!

**Comment by matbesancon** [comment]
@avik-pal was everything taken into account from this issue?

**Comment by avik-pal** [comment]
Yes. The last revision had all the changes.

**Comment by matbesancon** [comment]
Ok, I think you can close it then.
